quickwit/quickwit-control-plane/src/control_plane.rs

```diff
@@ -862,20 +859,46 @@ impl Handler<IndexerLeft> for ControlPlane {
     async fn handle(
         &mut self,
         message: IndexerLeft,
-        _ctx: &ActorContext<Self>,
+        ctx: &ActorContext<Self>,
     ) -> Result<Self::Reply, ActorExitStatus> {
         info!(
             "indexer `{}` left the cluster: rebuilding indexing plan",
             message.0.node_id()
         );
         // 1. Update shard table.
-        // 2. Rebalance shards if necessary.
-        self.ingest_controller.rebalance_shards();
+        self.ingest_controller
+            .rebalance_shards(&mut self.model, ctx.mailbox(), ctx.progress())
+            .await;
         self.indexing_scheduler.rebuild_plan(&self.model);
         Ok(())
     }
 }
 
+#[async_trait]
+impl Handler<RebalanceShardsCallback> for ControlPlane {
+    type Reply = ();
+
+    async fn handle(
+        &mut self,
+        message: RebalanceShardsCallback,
+        _ctx: &ActorContext<Self>,
+    ) -> Result<Self::Reply, ActorExitStatus> {
+        info!(
+            "closing {} shards after rebalance",
```

Comment:
I think we already have a code path to learn about shards being close:
router will tell use after getting bounced.

I don't think closing proactively like is preventing the other path, so I am not sure this message/handle is useful.

Reply:
Routers should not get bounced and ask the control plane for open shard alternatives because they should already know about the newly opened shards via gossip.

The other path that would let us know about those closed shards would be via ingesters broadcasting the status of their shards. However, this is propagated with some delay, which will cause the next rebalance operation to keep moving shards unnecessarily. We could also keep the rebalance lock for a few broadcast intervals, but I find that closing shards the current way is the simplest solution.